docs/generic-interfaces-infrastructure.md

```diff
@@ -0,0 +1,423 @@
+
+We can now move on to the actual CCW implementation. As with all interfaces, we'll need a vtable:
+
+```csharp
+public unsafe struct IEnumerator_stringVftbl
+{
+    public delegate* unmanaged[MemberFunction]<void*, Guid*, void**, HRESULT> QueryInterface;
+    public delegate* unmanaged[MemberFunction]<void*, uint> AddRef;
+    public delegate* unmanaged[MemberFunction]<void*, uint> Release;
+    public delegate* unmanaged[MemberFunction]<void*, uint*, Guid**, HRESULT> GetIids;
+    public delegate* unmanaged[MemberFunction]<void*, HSTRING*, HRESULT> GetRuntimeClassName;
+    public delegate* unmanaged[MemberFunction]<void*, TrustLevel*, HRESULT> GetTrustLevel;
+    public delegate* unmanaged[MemberFunction]<void*, void**, HRESULT> get_Current;
+    public delegate* unmanaged[MemberFunction]<void*, bool*, HRESULT> get_HasCurrent;
+    public delegate* unmanaged[MemberFunction]<void*, bool*, HRESULT> MoveNext;
+    public delegate* unmanaged[MemberFunction]<void*, int, void**, uint*, HRESULT> GetMany;
+}
+```
+
+The interop .dll can share a vtable implementation for compatible ABI types, to minimize codegen (eg. for all reference types).
+
+Now the actual CCW implementation type can be defined:
+
+```csharp
+public static unsafe class IEnumerator_stringImpl
+{
+    [FixedAddressValueType]
+    private static readonly IEnumerator_stringVftbl Vftbl;
+
+    static IEnumerator_stringImpl()
+    {
+        *(IInspectableVftbl*)Unsafe.AsPointer(ref Vftbl) = *(IInspectableVftbl*)IInspectableImpl.Vtable;
+
+        Vftbl.get_Current = &get_Current;
+        // Initialize the other vtable slots...
+    }
+
+    public static ref readonly Guid IID
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]   
+        get
+        {
+            ReadOnlySpan<byte> data = new byte[]
+            {
+                0x94, 0x2B, 0xEA, 0x94,
+                0xCC, 0xE9,
+                0xE0, 0x49,
+                0xC0,
+                0xFF,
+                0xEE,
+                0x64,
+                0xCA,
+                0x8F,
+                0x5B,
+                0x90
+            };
+
+            return ref Unsafe.As<byte, Guid>(ref MemoryMarshal.GetReference(data));
+        }
+    }
+
+    public static nint Vtable
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        get => (nint)Unsafe.AsPointer(ref Unsafe.AsRef(in Vftbl));
+    }
+
+    [UnmanagedCallersOnly(CallConvs = [typeof(CallConvMemberFunction)])]
+    private static HRESULT get_Current(void* thisPtr, bool* result)
+    {
+        try
+        {
+            var unboxedValue = ComInterfaceDispatch.GetInstance<IEnumerator<string>>((ComInterfaceDispatch*)thisPtr);
+
+            *result = IEnumerator_stringImplAdapterTable.GetAdapter(unboxedValue).get_Current();
+
+            return WellKnownErrorCodes.S_OK;
+        }
+        catch (global::System.Exception ex)
+        {
+            return RestrictedErrorInfoExceptionMarshaller.ConvertToUnmanaged(ex);
```

Comment:
Now that we have the new exception marshalling support (just added), you should be able to switch these implementations to be based on `[GeneratedComInterface]` if you wanted to do so.

Reply:
I believe this code should be generated in the sidecar? In that case `[GeneratedComInterface]` won't be available.

Reply:
Yeah this is all code in the interop .dll, so we can't leverage the COM generators here 😅
However that will be very useful for custom mixed classic COM scenarios!

Reply:
Right, this is just a C# representation of the IL you're generating into the sidecar.

Would be worth explicitly stating that somewhere in the doc if it isn't already.